Add support for custom resin buttons

luaneko opened the pull request:
Resolves #84 

This adds support for an arbitrary number of custom add/subtract buttons for the resin calculator. For example, buttons customized like this:

![image](https://user-images.githubusercontent.com/27663662/144735623-4518c34e-b609-4ef8-a57b-bcee02995d2d.png)

will appear like this:

![image](https://user-images.githubusercontent.com/27663662/144735647-c0b759ad-7835-4dbe-b2e0-a1eede493342.png)

Commit:
Add support for custom resin calculator buttons

Commit:
Fix resins spent statistics possibly becoming negative by adding resins over the limit

bschouwerwou commented on web/components/Home/Resin/Subtractor.tsx:
```diff
@@ -41,7 +42,7 @@ const Subtractor = ({ current }: { current: number }) => {
 const SubtractButton = ({ value, onClick }: { value: number; onClick: () => void }) => {
   const { formatMessage } = useIntl();
-  const hotkey = value.toString().slice(0, 1);
+  const hotkey = Math.abs(value).toString().slice(0, 1);
```
If +60 and -60 buttons are both shown, pressing 6 will trigger both and no change will occur.

Should we only support the hotkey for subtractions to keep the behaviour the same as before and prevent that issue? Or maybe use a combination hotkey for additions, like `shift+number`?

luaneko replied:
> Or maybe use a combination hotkey for additions, like `shift+number`?

I think this is a good approach since it's backwards compatible, while also allowing additions with keyboard shortcuts

bschouwerwou commented on web/components/Home/Resin/Subtractor.tsx:
```diff
@@ -1,36 +1,37 @@
+      {buttons.map(
         (value) =>
-          current - value > 0 && (
+          ((value < 0 && current + value >= 0) || (value > 0 && current < ResinCap)) && (
```
So if I've got 120 Resin and a +60 button, this logic will show the button, but pressing it would only result in +40 Resin because we enforce the 160 cap.

I don't think Genshin prevents you from going over the 160 Resin cap when topping up, but I'm not sure why a player would do that so I don't think it's super important for us to also support that.

However, if we don't let the user's Resin value go over 160 then I think we should change the condition for addition buttons so they're only shown if it can be added fully without going over the cap.

So that could be something like `value > 0 && ResinCap - roundResin(current) >= value`.

luaneko replied:
> I don't think Genshin prevents you from going over the 160 Resin cap when topping up

That is correct. That's why I didn't prevent the button from showing when the resin could overflow the limit.

> I'm not sure why a player would do that so I don't think it's super important for us to also support that.

That's fair.

bschouwerwou commented on web/components/Home/Resin/Subtractor.tsx:
```diff
@@ -1,36 +1,37 @@
 
-const resinUsages = [60, 40, 30, 20];
-
 const Subtractor = ({ current }: { current: number }) => {
```
Should we rename the `Subtractor` and `SubtractorButton` components since they don't exclusively subtract now?

Maybe `ResinCalcButtons`/`ResinCalcButton`?

luaneko replied:
I think we can simply call it `SideButton` since the file is already contained in the resin calculator folder. I'm surprised I didn't think about this.

Commit:
Merge branch 'master' into custom-resin-buttons

Commit:
Prevent potential shortcut conflicts in resin calculator

This keeps the resins subtraction convenient by pressing the number keys
while making additions require the shift modifier key.

Commit:
Update shortcut help window to reflect the new resin calculator keys bindings

Commit:
Prevent potential resin overflow when adding resins using side buttons

Commit:
Rename resin Subtractor to SideButtons

These buttons can add now, so it makes no sense to call it subtractors.

Commit:
Ensure resin calculator buttons are somewhat sane

Primajin commented:
When I manually add 60 and then substract it again, the statsitics on resins spent somehow misses that - does using that button solve that as well?

luaneko commented:
> When I manually add 60 and then substract it again, the statsitics on resins spent somehow misses that - does using that button solve that as well?

Yes. When the input is edited manually, the edited difference is subtracted from "resins spent" statistics. It behaves that way because it is intended to undo the mistake of subtracting too much. The new addition buttons added in this PR don't subtract from "resins spent" statistics because they are specifically intended to be used for resin gains like Fragile Resins.

Commit:
Update changelog for the addition of custom resin buttons

Primajin commented:
Ooohh perfect, thank you so much! 🚀
